Move expiry action and make probability of triggering it configurable

MKodde opened the pull request:
The State Storage PDO implementation was extended with a new configuration option, making it possible to set the probability the expired state storage is cleaned up.

And the expiry action was moved from the getValue method (called often) to the setValue method (called less often).

An effort was taken to improve test coverage. And at least cover the new features in test. @pmeulen already provided test coverage for the Factory tasked with creating the PDO state storage instance.

https://www.pivotaltracker.com/story/show/181553917

MKodde committed:
Install Mockery for testing purposes

MKodde committed:
Refactor PDO StateStorage

Goal is to add an additional configuration parameter (the probability
that the expired state is cleaned up) But other improvements have been
applied that where encountered and had little or no impact but
significantly improved the class.

1. Made the table a specific parameter instead of grabbin it from the
   'options' array
2. Did the same for the 'cleanupProbability' configuration as per [1]
   (task 2)
3. Moved the cleanup of expiry data to the 'setValue' method as per [1]
   (task 1)

[1] https://www.pivotaltracker.com/story/show/181553917

MKodde committed:
Integrate StateStorage PDO changes in factory

The factory now has to unpack the options, and verify the correct
options are present in the configuration. If they are not, a Runtime
exception is thrown. This can later be catched, or not if we want this
to halt the execution of the code path.

pmeulen commented on library/tiqr/tests/Tiqr_StateStoragePdoTest.php:
```diff
@@ -0,0 +1,61 @@
+        $this->stateStorage->setValue($key, random_bytes(5), 1);
+
+        $value = $this->stateStorage->getValue($key);
+        $this->stateStorage->shouldHaveReceived('cleanExpired');
```
Effective way to test that cleanup was called. Nice!

MKodde replied:
Wait till you see the version that actually worked ;)

pmeulen commented on library/tiqr/Tiqr/StateStorage/Pdo.php:
```diff
@@ -25,17 +25,39 @@
+     * @var int
+     */
+    private $cleanupProbability;
+
```
My suggestion would be to specify cleanupProbability as a float with allowed values 0 >= cleanupProbability >= 1. This is also the way probability is defined in statistics:
0: Cleanup will never run
0.2: Cleanup will run in 20% of the calls to setValue
1: Cleanup will always run

Using "1000" limits the resolution of the current implementation and binds cleanupProbability to this choice making it much harder to change this later, so instead I'd do (pseudocode):

```
if ( $cleanupProbability > (rand() / getrandmax()) ) {
   cleanup()
}
```

MKodde replied:
Yup! that makes way more sense.

MKodde committed:
Change probability calculating algorithm

Instead of testing on a range of 0..1000. We now set a probability based
on a percentage expressed in a floating point value between zero and
one. Where 0 is 0% and 1 expresses 100%.

This makes configuration more predictable.

Thanks for your suggestion and code contribution @pmeulen:
https://github.com/SURFnet/tiqr-server-libphp/pull/25/files#r837431802

VadimSchmitz reviewed (approved):
Looks good to me. The new probability implementation is definitely better.
